packages/toc/src/toc_item.tsx

```diff
@@ -2,7 +2,121 @@
 // Distributed under the terms of the Modified BSD License.
 
 import * as React from 'react';
-import { IHeading } from './utils/headings';
+import { CommandRegistry } from '@lumino/commands';
+import { IHeading, INotebookHeading } from './utils/headings';
+import { Menu } from '@lumino/widgets';
+import { CodeCell } from '@jupyterlab/cells';
+import { NotebookPanel } from '@jupyterlab/notebook';
+
+/**
+ * The command IDs used by TOC item.
+ */
+namespace CommandIDs {
+  export const runCells = 'toc:run-cells';
+}
+
+/**
+ * Adds commands to command registry and add items to context menu.
+ *
+ * @param contextMenu - context menu
+ * @param commands - command registry
+ * @param callbacks - Map which maps commandID with callback
+ * @returns
+ */
+const addCommands = (
+  contextMenu: Menu,
+  commands: CommandRegistry,
+  callbacks: Map<string, CommandRegistry.CommandFunc<any>>
+) => {
+  commands.addCommand(CommandIDs.runCells, {
+    execute: (callbacks.has(CommandIDs.runCells)
+      ? callbacks.get(CommandIDs.runCells)
+      : null) as CommandRegistry.CommandFunc<any>,
+    label: 'Run Cell(s)'
+  });
+  contextMenu.addItem({ command: CommandIDs.runCells });
```

Comment:
An alternative approach to re-defining the commands over and over (currently it would be redefined at each `render()` call) is to pass the callback to `args` which will be then passed on to the command's `execute`. For example, the open file command is added to context menu like this:

https://github.com/jupyterlab/jupyterlab/blob/4066cfe17f00db956ea91234b0fb67127acf5673/packages/filebrowser-extension/src/index.ts#L576-L580

and then then the factory gets extracted from the args in the `execute` here:

https://github.com/jupyterlab/jupyterlab/blob/4066cfe17f00db956ea91234b0fb67127acf5673/packages/filebrowser-extension/src/index.ts#L857-L860

I think that it would also make the code a bit more straightforward.